File: apps/fsc_custom/fsc_custom/public/js/sales_invoice_passthrough.js
// Copyright (c) 2026, FSC and contributors
// Client Script for Sales Invoice - Pass-Through Item Functionality

frappe.ui.form.on('Sales Invoice Item', {
    is_passthrough_item: function (frm, cdt, cdn) {
        let row = locals[cdt][cdn];

        if (row.is_passthrough_item) {
            // Set default buying price list if not set
            if (!row.buying_price_list) {
                frappe.model.set_value(cdt, cdn, 'buying_price_list', 'Standard Buying');
            }

            // Warn if update_stock is enabled
            if (row.update_stock == 1) {
                frappe.msgprint({
                    title: __('Stock Update'),
                    message: __('Pass-through items should not update stock. Please uncheck "Update Stock" for this item.'),
                    indicator: 'orange'
                });
            }
        } else {
            // Clear pass-through fields
            frappe.model.set_value(cdt, cdn, 'supplier', '');
            frappe.model.set_value(cdt, cdn, 'supplier_rate', 0);
            frappe.model.set_value(cdt, cdn, 'supplier_amount', 0);
            frappe.model.set_value(cdt, cdn, 'margin_amount', 0);
            frappe.model.set_value(cdt, cdn, 'margin_percentage', 0);
        }
    },

    supplier: function (frm, cdt, cdn) {
        let row = locals[cdt][cdn];

        if (row.is_passthrough_item && row.supplier && row.item_code && row.buying_price_list) {
            // Fetch supplier rate from buying price list
            fetch_supplier_rate(frm, cdt, cdn);
        }
    },

    buying_price_list: function (frm, cdt, cdn) {
        let row = locals[cdt][cdn];

        if (row.is_passthrough_item && row.supplier && row.item_code && row.buying_price_list) {
            // Fetch supplier rate from buying price list
            fetch_supplier_rate(frm, cdt, cdn);
        }
    },

    supplier_rate: function (frm, cdt, cdn) {
        calculate_passthrough_margin(frm, cdt, cdn);
    },

    qty: function (frm, cdt, cdn) {
        let row = locals[cdt][cdn];
        if (row.is_passthrough_item) {
            calculate_passthrough_margin(frm, cdt, cdn);
        }
    },

    rate: function (frm, cdt, cdn) {
        let row = locals[cdt][cdn];
        if (row.is_passthrough_item) {
            calculate_passthrough_margin(frm, cdt, cdn);
        }
    },

    amount: function (frm, cdt, cdn) {
        let row = locals[cdt][cdn];
        if (row.is_passthrough_item) {
            calculate_passthrough_margin(frm, cdt, cdn);
        }
    }
});

frappe.ui.form.on('Sales Invoice', {
    refresh: function (frm) {
        // Add pass-through summary to dashboard
        if (frm.doc.docstatus === 0 && frm.doc.items && frm.doc.items.length > 0) {
            show_passthrough_summary(frm);
        }

        // Add custom button to view linked Purchase Invoices
        if (frm.doc.docstatus === 1) {
            let has_passthrough = frm.doc.items.some(item => item.is_passthrough_item);

            if (has_passthrough) {
                frm.add_custom_button(__('View Purchase Invoices'), function () {
                    view_linked_purchase_invoices(frm);
                }, __('Pass-Through'));

                frm.add_custom_button(__('Passthrough Summary'), function () {
                    show_passthrough_detail_dialog(frm);
                }, __('Pass-Through'));
            }
        }
    },

    before_submit: function (frm) {
        // Validate pass-through items before submission
        return validate_passthrough_items(frm);
    }
});

function fetch_supplier_rate(frm, cdt, cdn) {
    let row = locals[cdt][cdn];

    frappe.call({
        method: 'erpnext.stock.get_item_details.get_price_list_rate',
        args: {
            args: {
                item_code: row.item_code,
                price_list: row.buying_price_list,
                supplier: row.supplier,
                company: frm.doc.company,
                qty: row.qty || 1,
                transaction_date: frm.doc.posting_date,
                doctype: 'Purchase Invoice'
            }
        },
        callback: function (r) {
            if (r.message && r.message.price_list_rate) {
                frappe.model.set_value(cdt, cdn, 'supplier_rate', r.message.price_list_rate);
                frappe.show_alert({
                    message: __('Supplier rate fetched: {0}', [format_currency(r.message.price_list_rate, frm.doc.currency)]),
                    indicator: 'green'
                });
            } else {
                frappe.msgprint({
                    title: __('Price Not Found'),
                    message: __('No price found for item {0} in price list {1} for supplier {2}',
                        [row.item_code, row.buying_price_list, row.supplier]),
                    indicator: 'orange'
                });
            }
        }
    });
}

function calculate_passthrough_margin(frm, cdt, cdn) {
    let row = locals[cdt][cdn];

    if (row.is_passthrough_item && row.supplier_rate) {
        // Calculate supplier amount
        row.supplier_amount = flt(row.qty) * flt(row.supplier_rate);

        // Calculate margin
        row.margin_amount = flt(row.amount) - flt(row.supplier_amount);

        // Calculate margin percentage (on cost basis)
        row.margin_percentage = row.supplier_amount ?
            (flt(row.margin_amount) / flt(row.supplier_amount) * 100) : 0;

        // Refresh the row
        frm.refresh_field('items');

        // Show warning if negative margin
        if (row.margin_amount < 0) {
            frappe.show_alert({
                message: __('Warning: Negative margin for item {0}!', [row.item_name]),
                indicator: 'red'
            }, 5);
        }
    }
}

function show_passthrough_summary(frm) {
    let passthrough_items = frm.doc.items.filter(item => item.is_passthrough_item);

    if (passthrough_items.length > 0) {
        let total_supplier_cost = 0;
        let total_revenue = 0;
        let suppliers = new Set();

        passthrough_items.forEach(item => {
            total_supplier_cost += flt(item.supplier_amount);
            total_revenue += flt(item.amount);
            if (item.supplier) {
                suppliers.add(item.supplier);
            }
        });

        let total_margin = total_revenue - total_supplier_cost;
        let margin_pct = total_supplier_cost ? (total_margin / total_supplier_cost * 100) : 0;

        frm.dashboard.add_indicator(__('Pass-Through Items: {0}', [passthrough_items.length]), 'blue');
        frm.dashboard.add_indicator(__('Suppliers: {0}', [suppliers.size]), 'blue');
        frm.dashboard.add_indicator(__('Total Margin: {0} ({1}%)',
            [format_currency(total_margin, frm.doc.currency), margin_pct.toFixed(2)]),
            total_margin >= 0 ? 'green' : 'red'
        );
    }
}

function validate_passthrough_items(frm) {
    let errors = [];

    frm.doc.items.forEach((item, idx) => {
        if (item.is_passthrough_item) {
            // Check supplier is set
            if (!item.supplier) {
                errors.push(__('Row {0}: Supplier is required for pass-through item', [idx + 1]));
            }

            // Check supplier rate is set
            if (!item.supplier_rate || item.supplier_rate <= 0) {
                errors.push(__('Row {0}: Supplier rate must be greater than 0', [idx + 1]));
            }

            // Warn on negative margin
            if (item.margin_amount < 0) {
                errors.push(__('Row {0}: Warning - Negative margin ({1})',
                    [idx + 1, format_currency(item.margin_amount, frm.doc.currency)]));
            }

            // Check expense account is set
            if (!item.expense_account) {
                errors.push(__('Row {0}: Expense account is required for pass-through item', [idx + 1]));
            }
        }
    });

    if (errors.length > 0) {
        frappe.msgprint({
            title: __('Pass-Through Validation Errors'),
            message: errors.join('<br>'),
            indicator: 'red'
        });
        return false;
    }

    return true;
}

function view_linked_purchase_invoices(frm) {
    frappe.route_options = {
        "custom_sales_invoice": frm.doc.name
    };
    frappe.set_route("List", "Purchase Invoice");
}

function show_passthrough_detail_dialog(frm) {
    let passthrough_items = frm.doc.items.filter(item => item.is_passthrough_item);

    if (passthrough_items.length === 0) {
        frappe.msgprint(__('No pass-through items in this invoice'));
        return;
    }

    // Group by supplier
    let supplier_groups = {};
    passthrough_items.forEach(item => {
        if (!supplier_groups[item.supplier]) {
            supplier_groups[item.supplier] = [];
        }
        supplier_groups[item.supplier].push(item);
    });

    // Build HTML table
    let html = '<table class="table table-bordered table-sm">';
    html += '<thead><tr><th>Supplier</th><th>Items</th><th>Supplier Cost</th><th>Revenue</th><th>Margin</th><th>Margin %</th><th>PI</th></tr></thead>';
    html += '<tbody>';

    for (let supplier in supplier_groups) {
        let items = supplier_groups[supplier];
        let total_cost = items.reduce((sum, item) => sum + flt(item.supplier_amount), 0);
        let total_revenue = items.reduce((sum, item) => sum + flt(item.amount), 0);
        let margin = total_revenue - total_cost;
        let margin_pct = total_cost ? (margin / total_cost * 100) : 0;
        let pi_link = items[0].purchase_invoice ?
            `<a href="/app/purchase-invoice/${items[0].purchase_invoice}">${items[0].purchase_invoice}</a>` :
            'Not Created';

        html += '<tr>';
        html += `<td><strong>${supplier}</strong></td>`;
        html += `<td>${items.length}</td>`;
        html += `<td>${format_currency(total_cost, frm.doc.currency)}</td>`;
        html += `<td>${format_currency(total_revenue, frm.doc.currency)}</td>`;
        html += `<td style="color: ${margin >= 0 ? 'green' : 'red'}">${format_currency(margin, frm.doc.currency)}</td>`;
        html += `<td>${margin_pct.toFixed(2)}%</td>`;
        html += `<td>${pi_link}</td>`;
        html += '</tr>';
    }

    html += '</tbody></table>';

    frappe.msgprint({
        title: __('Pass-Through Details'),
        message: html,
        wide: true
    });
}

File: apps/fsc_custom/fsc_custom/sales_invoice_passthrough.py
# Copyright (c) 2026, FSC and contributors
# Server-side logic for Sales Invoice Pass-Through functionality

import frappe
from frappe import _
from frappe.utils import flt, getdate


def on_submit(doc, method):
	"""Create Purchase Invoices for pass-through items on Sales Invoice submission"""
	create_purchase_invoices_for_passthrough(doc)


def on_cancel(doc, method):
	"""Cancel linked Purchase Invoices when Sales Invoice is cancelled"""
	cancel_linked_purchase_invoices(doc)


def update_stock_ledger(doc, method):
	"""Override stock ledger updates to exclude pass-through items"""
	# This hook prevents stock updates for pass-through items
	# Implementation depends on how ERPNext processes stock
	pass


def create_purchase_invoices_for_passthrough(sales_invoice):
	"""
	Group pass-through items by supplier and create Purchase Invoices
	"""
	# Group items by supplier
	supplier_items = {}
	
	for item in sales_invoice.items:
		if item.is_passthrough_item and item.supplier:
			# Validate required fields
			if not item.supplier_rate or item.supplier_rate <= 0:
				frappe.throw(_(f"Row {item.idx}: Supplier rate is required for pass-through item {item.item_code}"))
			
			if not item.expense_account:
				frappe.throw(_(f"Row {item.idx}: Expense account is required for pass-through item {item.item_code}"))
			
			# Group by supplier
			if item.supplier not in supplier_items:
				supplier_items[item.supplier] = []
			supplier_items[item.supplier].append(item)
	
	# Create Purchase Invoice for each supplier
	purchase_invoices = []
	
	for supplier, items in supplier_items.items():
		try:
			pi = create_purchase_invoice(sales_invoice, supplier, items)
			purchase_invoices.append(pi.name)
			
			# Update Sales Invoice items with PI reference
			for item in items:
				frappe.db.set_value("Sales Invoice Item", item.name, "purchase_invoice", pi.name, update_modified=False)
			
			frappe.msgprint(
				_("Purchase Invoice {0} created for supplier {1}").format(
					f'<a href="/app/purchase-invoice/{pi.name}">{pi.name}</a>', 
					supplier
				),
				indicator="green",
				alert=True
			)
			
		except Exception as e:
			frappe.log_error(f"Error creating Purchase Invoice for supplier {supplier}: {str(e)}")
			frappe.throw(_(f"Failed to create Purchase Invoice for supplier {supplier}: {str(e)}"))
	
	# Commit the PI references
	frappe.db.commit()
	
	return purchase_invoices


def create_purchase_invoice(sales_invoice, supplier, items):
	"""Create a single Purchase Invoice for a supplier"""
	
	pi = frappe.new_doc("Purchase Invoice")
	
	# Set header fields
	pi.supplier = supplier
	pi.company = sales_invoice.company
	pi.posting_date = sales_invoice.posting_date
	pi.set_posting_time = 1
	pi.posting_time = sales_invoice.posting_time
	pi.currency = sales_invoice.currency
	pi.conversion_rate = sales_invoice.conversion_rate or 1
	pi.buying_price_list = items[0].buying_price_list if items else "Standard Buying"
	
	# Link back to Sales Invoice
	pi.custom_sales_invoice = sales_invoice.name
	
	# Set remarks
	pi.remarks = _(f"Auto-created from Sales Invoice {sales_invoice.name} for pass-through items")
	
	# Add items
	for si_item in items:
		pi.append("items", {
			"item_code": si_item.item_code,
			"item_name": si_item.item_name,
			"description": si_item.description or si_item.item_name,
			"qty": si_item.qty,
			"uom": si_item.uom,
			"stock_uom": si_item.stock_uom,
			"conversion_factor": si_item.conversion_factor or 1,
			"rate": si_item.supplier_rate,
			"amount": si_item.supplier_amount,
			"expense_account": si_item.expense_account,
			"cost_center": si_item.cost_center,
			"project": si_item.project if hasattr(si_item, 'project') else None,
			# Important: Don't update stock for pass-through items
			"update_stock": 0
		})
	
	# Calculate totals
	pi.run_method("calculate_taxes_and_totals")
	
	# Insert without submit (leave as Draft for review)
	# Users can review and submit manually
	pi.flags.ignore_permissions = True
	pi.insert()
	
	# Optional: Auto-submit if configured
	# Uncomment the next two lines to auto-submit
	# pi.submit()
	# frappe.msgprint(_(f"Purchase Invoice {pi.name} submitted"), indicator="green")
	
	return pi


def cancel_linked_purchase_invoices(sales_invoice):
	"""Cancel all Purchase Invoices linked to this Sales Invoice"""
	
	# Find all linked Purchase Invoices
	linked_pis = frappe.db.sql("""
		SELECT DISTINCT name
		FROM `tabPurchase Invoice`
		WHERE custom_sales_invoice = %s
		AND docstatus = 1
	""", sales_invoice.name, as_dict=1)
	
	for pi_row in linked_pis:
		try:
			pi = frappe.get_doc("Purchase Invoice", pi_row.name)
			pi.flags.ignore_permissions = True
			pi.cancel()
			
			frappe.msgprint(
				_("Purchase Invoice {0} cancelled").format(pi.name),
				indicator="orange"
			)
			
		except Exception as e:
			frappe.log_error(f"Error cancelling Purchase Invoice {pi_row.name}: {str(e)}")
			# Don't throw error, just log it - allow SI cancellation to proceed
			frappe.msgprint(
				_("Warning: Could not cancel Purchase Invoice {0}: {1}").format(pi_row.name, str(e)),
				indicator="red"
			)
	
	# Also cancel draft PIs (remove reference but don't cancel since they're not submitted)
	draft_pis = frappe.db.sql("""
		SELECT DISTINCT name
		FROM `tabPurchase Invoice`
		WHERE custom_sales_invoice = %s
		AND docstatus = 0
	""", sales_invoice.name, as_dict=1)
	
	for pi_row in draft_pis:
		try:
			frappe.delete_doc("Purchase Invoice", pi_row.name, force=1)
			frappe.msgprint(
				_("Draft Purchase Invoice {0} deleted").format(pi_row.name),
				indicator="orange"
			)
		except Exception as e:
			frappe.log_error(f"Error deleting draft Purchase Invoice {pi_row.name}: {str(e)}")


@frappe.whitelist()
def get_passthrough_summary(sales_invoice_name):
	"""Get summary of pass-through items in a Sales Invoice"""
	
	doc = frappe.get_doc("Sales Invoice", sales_invoice_name)
	
	passthrough_items = [item for item in doc.items if item.is_passthrough_item]
	
	if not passthrough_items:
		return {
			"total_items": 0,
			"total_suppliers": 0,
			"total_cost": 0,
			"total_revenue": 0,
			"total_margin": 0,
			"margin_percentage": 0,
			"supplier_breakdown": {}
		}
	
	# Group by supplier
	supplier_groups = {}
	for item in passthrough_items:
		if item.supplier not in supplier_groups:
			supplier_groups[item.supplier] = {
				"items": [],
				"total_cost": 0,
				"total_revenue": 0,
				"total_margin": 0,
				"purchase_invoice": item.purchase_invoice
			}
		
		supplier_groups[item.supplier]["items"].append({
			"item_code": item.item_code,
			"item_name": item.item_name,
			"qty": item.qty,
			"supplier_rate": item.supplier_rate,
			"supplier_amount": item.supplier_amount,
			"rate": item.rate,
			"amount": item.amount,
			"margin_amount": item.margin_amount,
			"margin_percentage": item.margin_percentage
		})
		
		supplier_groups[item.supplier]["total_cost"] += flt(item.supplier_amount)
		supplier_groups[item.supplier]["total_revenue"] += flt(item.amount)
		supplier_groups[item.supplier]["total_margin"] += flt(item.margin_amount)
	
	# Calculate totals
	total_cost = sum(flt(item.supplier_amount) for item in passthrough_items)
	total_revenue = sum(flt(item.amount) for item in passthrough_items)
	total_margin = total_revenue - total_cost
	margin_percentage = (total_margin / total_cost * 100) if total_cost else 0
	
	return {
		"total_items": len(passthrough_items),
		"total_suppliers": len(supplier_groups),
		"total_cost": total_cost,
		"total_revenue": total_revenue,
		"total_margin": total_margin,
		"margin_percentage": margin_percentage,
		"supplier_breakdown": supplier_groups
	}
